src/main/java/org/javarosa/core/model/ItemsetBinding.java

```diff
@@ -200,53 +201,60 @@ private SelectChoice getChoiceForTreeReference(FormDef formDef, DataInstance for
+        IAnswerData originalValue = formDef.getMainInstance().resolveReference(curQRef).getValue();
 
-        if (currentAnswersInNewChoices != null && currentAnswersInNewChoices.containsValue(false)) {
-            IAnswerData filteredAnswer;
-            if (rawValue instanceof MultipleItemsData) {
-                filteredAnswer = getFilteredSelections((MultipleItemsData) rawValue, currentAnswersInNewChoices);
+        if (currentAnswersInNewChoices != null) {
+            IAnswerData boundAndFilteredValue;
+            if (originalValue instanceof MultipleItemsData) {
+                boundAndFilteredValue = getFilteredAndBoundSelections((MultipleItemsData) originalValue, currentAnswersInNewChoices);
+            } else if (currentAnswersInNewChoices.containsValue(null)) {
+                boundAndFilteredValue = new StringData("");
             } else {
-                filteredAnswer = new StringData("");
+                SelectChoice selectChoice = currentAnswersInNewChoices.get(originalValue.getDisplayText());
+                boundAndFilteredValue = new SelectOneData(selectChoice.selection());
             }
 
-            formDef.getMainInstance().resolveReference(curQRef).setAnswer(filteredAnswer);
+            formDef.getMainInstance().resolveReference(curQRef).setAnswer(boundAndFilteredValue);
         }
     }
 
     /**
      * @param selections          an answer to a multiple selection question
-     * @param shouldKeepSelection maps each value that could be in @{code selections} to a boolean representing whether
-     *                            or not it should be kept
-     * @return a copy of {@code selections} without the values that were mapped to false in {@code shouldKeepSelection}
+     * @param selectChoicesToKeep maps each value that could be in @{code selections} to a SelectChoice if it should be bound
+     *                            or null if it should be removed.
+     * @return a copy of {@code selections} without the values that were mapped to false in {@code selectChoicesToKeep} and
+     * with all selections bound.
      */
-    private static MultipleItemsData getFilteredSelections(MultipleItemsData selections, Map<String, Boolean> shouldKeepSelection) {
+    private static MultipleItemsData getFilteredAndBoundSelections(MultipleItemsData selections, Map<String, SelectChoice> selectChoicesToKeep) {
         List<Selection> newSelections = new ArrayList<>();
         for (Selection oldSelection : (List<Selection>) selections.getValue()) {
             String key = oldSelection.choice != null ? oldSelection.choice.getValue() : oldSelection.xmlValue;
-            if (shouldKeepSelection.get(key)) {
-                newSelections.add(oldSelection);
+            if (selectChoicesToKeep.get(key) != null) {
+                newSelections.add(selectChoicesToKeep.get(key).selection());
```

Comment:
May I suggest
```
if (selectChoicesToKeep.containsKey(key)) {
    newSelections.add(selectChoicesToKeep.get(key).selection());
}
```

or

```
SelectChoice choice = selectChoicesToKeep.get(key);
if (choice != null) {
    newSelections.add(choice.selection());
}
```

Reply:
I think the first would crash without other changes because we are keeping keys that shouldn't exist anymore and mapping them to null. But maybe we could remove the keys instead -- will look at that.

Reply:
I explored keys instead of having them map to null and it's not practical. We currently initialize the map keys based on an old answer which could contain unavailable choices. All values are initially null. We then iterate over all the currently-available choices and set values for any choices that are in the old answer map. Removing keys mapped to null would have to be a separate step and it doesn't seem like it adds clarity. I went with your second suggestion.

While exploring this, I noticed some of my naming didn't make as much sense with the new mapping to `SelectChoice` objects (replacing booleans). I tried to improve that and add a few more comments where I had to do some extra thinking to figure out what was going on.